bundles/org.openhab.binding.groheondus/src/main/java/org/openhab/binding/groheondus/internal/handler/GroheOndusAccountHandler.java

```diff
@@ -77,13 +79,13 @@ public void deleteRefreshToken() {
 
     public void setRefreshToken(String refreshToken) {
         this.storage.put(STORAGE_KEY_REFRESH_TOKEN, refreshToken);
-        this.initialize();
+        this.login();
     }
 
     private void scheduleTokenRefresh() {
         if (ondusService != null) {
             Instant expiresAt = ondusService.authorizationExpiresAt();
-            Duration between = Duration.between(Instant.now(), expiresAt);
+            Duration between = Duration.between(Instant.now(), expiresAt.minus(2, ChronoUnit.MINUTES));
```

Comment:
any specific thinking why the token should be refreshed 2 minutes before it expires? Usually, I know that libraries tend to use a certain percentage, instead of a fixed time. Something like "refresh the token when 75% of it's lifetime passed). I'm okay with a fixed, one, as well. However, why 2 minutes :D Just out of interest :P

Reply:
No other reason that it needs to happen _before_ it expires. And the way it was before it could happen either a split second before - or after - rendering it unusable. Changed to 2 hours before, but really does not matter.

What I have found lately is that the token service also may fail, so I added error handling and re-login in this case.

A problem in the current impl is that the token refresh is living its own life apart from actual service calls. So in case the token refresh fails & a service call is about to happen at the same time, the service call may fail during the time it takes to issue a new token.